pkg/issuer/vault/issue.go

```diff
@@ -130,7 +132,80 @@ func (v *Vault) configureCertPool(cfg *vault.Config) error {
 	return nil
 }
 
+var (
+	clientRepo   map[*v1alpha1.VaultIssuer]*vault.Client
+	clientRepoMu sync.Mutex
+)
+
+func renewToken(vaultClient *vault.Client) {
+	resp, err := vaultClient.Auth().Token().LookupSelf()
+	if err != nil {
+		glog.Warningf("vault: lookup-self failed, token renewal is disabled: %s", err)
+		return
+	}
+
+	b, _ := json.Marshal(resp.Data)
+	var data struct {
+		TTL         int       `json:"ttl"`
+		CreationTTL int       `json:"creation_ttl"`
+		Renewable   bool      `json:"renewable"`
+		ExpireTime  time.Time `json:"expire_time"`
+	}
+	if err := json.Unmarshal(b, &data); err != nil {
+		glog.Warningf("vault: lookup-self failed, token renewal is disabled: %s", err)
+		return
+	}
+
+	switch {
+	case data.Renewable:
+		// renewable token, proceed further
+	case data.ExpireTime.IsZero():
+		// nothing to do. token doesn't expire
+		return
+	default:
+		ttl := time.Until(data.ExpireTime) / time.Second * time.Second // truncate to secs
+		glog.Warningf("vault: Token is not renewable and it will expire %s from now at %s",
+			ttl, data.ExpireTime.Format(time.RFC3339))
+		return
+	}
+
+	ttl := time.Duration(data.TTL) * time.Second
+	timer := time.NewTimer(ttl / 2)
+
+	for range timer.C {
```

Comment:
We'll leak goroutines here as we currently have no way to evict vault clients from the cache.

I think we need to think about how eviction works (i.e. stopping renewing and removing from cache) and then plumb that through to here so that we can return from this loop.

Reply:
When a token is no longer renewable or the loop failed to renew one, it clear the token from vault client and inside initVaultClient(), for a cached vault client if a token is not set, the client is deleted from the cache.

I am not following why we'll leak goroutines here as goroutines will return (that means it exits, yes) when they have problems renew tokens.

Reply:
Sure, but what if a user deletes the Issuer resource containing this configuration?

We can't just assume that all tokens will eventually expire/fail to renew, this isn't true in all cases and can cause leaks 😬

Reply:
I just added logic to check issuer updates.

Reply:
I added logic to check for issuer updates/deletions.